File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/collision/strategy/dual/BruteDualCollisionStrategy.java
package com.greentree.commons.geometry.geom2d.collision.strategy.dual;

import java.util.Collection;
import java.util.HashSet;

import com.greentree.commons.geometry.geom2d.collision.Collidable2D;
import com.greentree.commons.util.cortege.UnOrentetPair;

public class BruteDualCollisionStrategy<T extends Collidable2D>
		implements DualCollisionStrategy<T> {
	
	@Override
	public HashSet<UnOrentetPair<T>> getCollisionContact(Collection<T> ac, Collection<T> bc) {
		HashSet<UnOrentetPair<T>> res = new HashSet<>();
		int i = 0, j;
		for(var a : ac) {
			j = 0;
			for(var b : bc)
				if(i < j)
					break;
				else {
					if(a != b && a.getShape().isIntersect(b.getShape()))
						res.add(new UnOrentetPair<>(a, b));
					j++;
				}
			i++;
		}
		return res;
	}
	
}

File: Commons/Geometry/src/main/java/com/greentree/commons/geometry/geom2d/collision/strategy/dual/BruteDualCollisionStrategy.kt
package com.greentree.commons.geometry.geom2d.collision.strategy.dual

import com.greentree.commons.geometry.geom2d.collision.Collidable2D
import com.greentree.commons.geometry.geom2d.isIntersect
import com.greentree.commons.util.cortege.UnOrentetPair

class BruteDualCollisionStrategy<T : Collidable2D>
	: DualCollisionStrategy<T> {

	override fun getCollisionContact(ac: Collection<T>, bc: Collection<T>): HashSet<UnOrentetPair<T>> {
		val res = HashSet<UnOrentetPair<T>>()
		var i = 0
		var j: Int
		for(a in ac) {
			j = 0
			for(b in bc) if(i < j) break
			else {
				if(a !== b && a.shape.isIntersect(b.shape)) res.add(UnOrentetPair(a, b))
				j++
			}
			i++
		}
		return res
	}
}
